### ISSUE TITLE
Some tests fail with upcoming release of R6

### ISSUE THREAD

**Issue by wch** [hongyuanjia/eplusr#164]
I'm about to submit a new version of R6 to CRAN, but am getting errors with this package when I run checks on reverse dependencies.

Some background: In previous versions of R6, when cloning an object with active bindings, it will extract the function for the active binding at _cloning time_ and put it in the new object. This relied on behavior from `as.list.environment` which apparently was a bug, and will be changed in upcoming versions of R. The bug was that `as.list.environment` would give the function behind an active binding, rather than the value that the function returns when invoked. In upcoming versions of R, it will call the function and return that value.

I have been asked by a Luke Tierney (who is a member of R-Core) to release a new version of R6 which does not rely on the buggy behavior of `as.list.environment`, so that he can fix the bug in R. When the bug is fixed, it will no longer be possible to get the function for an active binding.

In the development version of R6, it separately stores the functions for active bindings when an R6 object is created, and uses those stored functions to create new active bindings when the object is cloned.

Currently, tests for the CRAN version of eplusr fail with the development version of R6.

<details>
  <summary>eplusr check log</summary>

```
Package: eplusr
Check: tests
New result: ERROR
    Running â€˜testthat.Râ€™ [87s/37s]
  Running the tests in â€˜tests/testthat.Râ€™ failed.
  Complete output:
    > library(testthat)
    > library(eplusr)
    > eplusr_option(verbose_info = FALSE)
    $verbose_info
    [1] FALSE
    
    > 
    > test_check("eplusr")
    trying URL 'https://raw.githubusercontent.com/NREL/EnergyPlus/v9.1.0/idd/V8-8-0-Energy%2B.idd'
    Content type 'text/plain; charset=utf-8' length 4055399 bytes (3.9 MB)
    ==================================================
    downloaded 3.9 MB
    
    â”€â”€ 1. Failure: Idf class (@test_idf.R#478)  â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
    idf2$Zone[[1]]$Name not equal to "ZONE ONE".
    1/1 mismatches
    x[1]: "zone"
    y[1]: "ZONE ONE"
    
    â”€â”€ 2. Failure: Insert (@test_impl-idf.R#805)  â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
    ins$object not equivalent to data.table(...).
    Different number of rows
    
    â”€â”€ 3. Failure: Insert (@test_impl-idf.R#810)  â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
    ins$value$value_id not equivalent to 349:357.
    Lengths differ: 0 is not 9
    
    â”€â”€ 4. Error: Insert (@test_impl-idf.R#811)  â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
    subscript out of bounds
    1: expect_equivalent(ins$value$value_chr[[1L]], "new_mat") at testthat/test_impl-idf.R:811
    2: quasi_label(enquo(object), label, arg = "object")
    3: eval_bare(get_expr(quo), get_env(quo))
    
    â•â• testthat results  â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•
    [ OK: 1096 | SKIPPED: 7 | WARNINGS: 0 | FAILED: 4 ]
    1. Failure: Idf class (@test_idf.R#478) 
    2. Failure: Insert (@test_impl-idf.R#805) 
    3. Failure: Insert (@test_impl-idf.R#810) 
    4. Error: Insert (@test_impl-idf.R#811) 
    
    Error: testthat unit tests failed
    Execution halted
```
</details>

Please install the development version of R6 and release a new version of eplusr which has passing tests.

```R
devtools::install_github('r-lib/R6@rc-v2.4.1')
```

In some cases, errors occur only when the package is loaded the normal way, but not when `load_all()` is used.

Since I have been asked by a member of R-core to submit this fixed version of R6, I want to release it as soon as possible, but this requires that an updated version of eplusr is on CRAN first. Thank you.

**Comment by hongyuanjia** [comment]
Thanks @wch! I will look into it and submit a new version as soon as possible.

**Comment by hongyuanjia** [comment]
@wch, I can locate where the problem comes from. However, I did not know how to handle it correctly. Here is the case:

* In my R6 class `Idf`, there is no active fields defined in the definitionusing `R6::R6Class`.
* After an `Idf` object is created, I manually create active bindings based on the `Idf` object contents using `base::makeActiveBinding()` in function [`add_idf_class_bindings()`](https://github.com/hongyuanjia/eplusr/blob/272a2a358377ba1b930d5dbec8b6382a63d2d91a/R/idf.R#L2660), in order to provide dynamic content query. The results of those queries are a list of R6 objects of another class called `IdfObject` and `IdfObject` does not have its own `deep_clone()` method. 
  https://github.com/hongyuanjia/eplusr/blob/272a2a358377ba1b930d5dbec8b6382a63d2d91a/R/idf.R#L2658-L2666

* I defined a custom `deep_clone()` method in `Idf` private environment, which is pretty simple. At that time, the purpose of this function is to create a copy of private fields which are environments containing multiple data.tables. I did not know that I have already depended on the buggy behavior of `as.list.environment()` to get the active binding function content...
  https://github.com/hongyuanjia/eplusr/blob/272a2a358377ba1b930d5dbec8b6382a63d2d91a/R/idf.R#L2477-L2488

* Now with the development version of R6, as these active bindings are not `active` fields defined in `R6::R6Class` and `.__enclos_env__$.__active__` being `NULL`, [`binding_names` here](https://github.com/r-lib/R6/blob/a0ef58573d0a780bb84836b799d85395be687045/R/clone.R#L216) will catch them and treat them as normal methods. After cloning, the binding environment of those active bindings will still be the old `Idf`, not the new one.

There are two ways coming to my mind:

1. Define a custom public `clone()` method that does the same as the default `clone()` method defined in R6 package, except that it excludes or directly removes all those active bindings that are not defined in this R6 class. The problem is that I did not know how to call the default `clone()` method in `R6::R6Class()`.

2. In the custom `deep_clone()` method, catch that situation and assign new enclosing environment of the new cloned object. However I did not know how to get access to the new enclosing environment in the original private environment when `deep_clone()` is called, except the hacking `parent.frame(3)`...

The first one is preferable. Could you please give me some advice on how I can provide a new public `clone()` method and at the same time call the default `clone()` method in that function? Thanks a lot!

**Comment by wch** [comment]
The `clone()` method is quite complicated and captures a lot of edge cases (most of the difficulty is related to inheritance). I would not recommend trying to copy and modify it.

There is a base R function called `bindingIsActive()` which can tell you whether a binding is an active one. I think you can use that in your `deep_clone()` method to avoid copying active bindings.

**Comment by hongyuanjia** [comment]
Thanks @wch.

I tried to adopt your suggestion and return `NULL` when it is an active binding. However after cloning, they are still in the new cloned object. As this issue seems urgent, I do not want to hold you back too long for submitting new R6 because of eplusr package. So I used the first dirty approach #165. The new version of eplusr has passed tests with `R6@rc-v2.4.1` and successfully landed on CRAN.

I will adopt the 2nd approach in the next release.
